### PR TITLE
s4696417: Stable Diffusion for ADNI Dataset

### PR DESCRIPTION
**Pull request by s1lentcs**
Hi Teaching Team,

This is my pull request for my implementation of a Stable Diffusion model for task 8 using PyTorch. This implementation uses a VAE to encode input images into a latent form and applying the forward and backward noise processes in this space. The UNet portion of the model is then used to learn to predict the level of noise in an image so that it can be removed in the process of image generation. A breakdown of the model components, training and generation samples are all included in the detailed report in the README. 

The project contains the following:
- `README.md` contains the report with detailed information on the model implementation, training, results and visusalisations including a TSNE embedding. 
- `train.py` contains the primary code used to train the StableDiffusion model which can then be used to generate new samples
- `pre_train.py` contains the code to train the VAE model on the ADNI dataset
- `predict.py` contains the required code to generate new sample images from a trained StableDiffusion model
- `vae.py` contains the implementation of the Variational Autoencoder used by the Stable Diffusion model. 
- `modules.py` contains all other component modules for the Stable Diffusion model as well as implementation of a warmup scheduler and noise scheduler. 
- `utils.py` contains utility functions for visualisations and various setup steps 
- `dataset.py` contains the custom dataset class for the ADNI dataset as well as a functions to get dataloaders for this dataset
- `tsne.py` contains the code required to generate a TSNE embedding plot for samples generrated from the AD/NC Diffusion models

Thanks,
Liam O'Sullivan
s4696417

### CODE REVIEW

**Comment by gayanku** [comment]
*****************************************************************
**This is an initial inspection, no action is required at this point**
Stable Diffusion ADNI brain data ---- Hard Difficulty
| Category	| | Marks | Comments |	
| :---  | :---: | :---: | :---  |		
| Algorithm solves the problem 		| 5 | 4 | Unconditional, but this is fine. However, how did you show the AD NC images separate? Also,  Images not too clear. |																	|
| Implementation functions as intended	| 3 | 2 | |
| Good design 					| 1 | 0.5 | Modular, somewhat reusable, avoid hardcoded values (0.18215 in VAE), or add some clear comments in place  |
| Commenting 					| 1 | 1 | Meaningful docstrings and comments |
| Algorithm above Normal Difficulty 	| 5 | 5 | |
| Algorithm is Hard difficulty 			| 5 | 5 | Hard Difficulty |
| Section IV : Max mark 20 from 20		|   | **18**   | |

**TSNE / UMAP:** 
No TSNE, but shows AD NC images separate. From predict.py its not clear how showing AD NC separate is possible.
While no marks have been reduced, the task is to train one model on the whole dataset, and by plotting the embedding of style code, it is expected to see somehow clustered patterns for different classes.

**Suggestions:**
You seem to be pretraining VAE at image_size=125, then using IMAGE_SIZE = 256 in the train, also commenting images sizes need to match. Not sure if this is an error.

**Comment by s1lentcs** [comment]
Hi Gayan, I am a little confused by this feedback. 

For the TSNE plot I sampled from two different models, one trained on the NC class and one on the AD class. This is what I was instructed to do from the tutor team on ED. Also not too sure what you mean by images not too clear? Comparing to images from other students I feel that mine are more clear than most others and are also at the original size of the input data. I would definitely say that the images generated by my model are reasonably clear at a minimum. 

I am unsure how this solution does not function as intended? The VAE value of 0.18215 is a constant well known among the literature and has been labeled accordingly in the code? 

In regards to the VAE image size, when training the final stable diffusion model, I used the VAE which was trained on images of size 256x256, this may not be what is shown as the parameters in the code though. I am not sure what you mean by no TSNE? The plot is clearly shown and labelled in the README. 

Could you please provide some feedback regarding some of these comments?

**Comment by gayanku** [comment]
@s1lentcs,  This is a typo, I've corrected it now. No marks has been deducted for the TSNE.  However, looking at your predict there is no way someone can understand how to use it to generate AD/NC. You could add some comments, clearly label which pre-trained/check point to use or give some indication.  Now, after you mentioned it, I see you have given some instructions in the readme. However, some notes/comments in the predict is needed.

Your images are a bit unclear, and there are hard coded values in the design which limits its reusability.  Also there is some lack of clarity on the image size.

**Comment by YashMitttal** [comment]
Some additional feedback:

- You should justify why you chose certain hyperparameter values.
- Result images are not very clear
- Your graphs have no axes labels


Overall, good job!

**Comment by s1lentcs** [comment]
Hi Team, can I confirm the image quality is being assessed based on the images shown in the inference section of my README? I am really not sure how these are not that clear? This is from an inference of the model. 
normal_sample

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Adequate design and implementation. Image quality low.</td><td>-1</td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>OK solution to problem. Hard coded values. Some inconsistencies in code parameters.</td><td>-1</td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard. Stable Diffusion (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Acceptable. Minor improvements needed.</td><td>-1</td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF submitted.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). On time submission.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-3</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by shakes76** [comment]
Good work!
